### PR TITLE
feat: session api refactor

### PR DESCRIPTION
**Pull request by jhgg**
re-opening #96

Internally, we've (discord) wrapped `cassandra-rs` to basically implicitly associate a given Statement, PreparedStatement & Batch with a given Session.

This PR tries to improve upon that, by building that betterment directly into the library. This is most definitely a breaking change, but I think for the better. Given the ubiquity of async rust, I've also dropped support for the `.wait()` function on `CassFuture`, and some functions which used to return a CassFuture, now are just simple `async fn`'s. If you are operating in a synchronous space, you can use something like https://docs.rs/futures/0.3.8/futures/executor/fn.block_on.html - which will provide the machinery required to execute the now async-only code.

Essentially, `Session` now implements `Clone`. The inner pointer to the cassandra session has been moved into a `SessionInner`, which is now wrapped in an `Arc` inside of `Session`.

From there, we can now `Clone` the given `Session` to the relevant `Statement`, `PreparedStatement`, and `Batch` (associated session resource).  This provides a more safe API, as it makes it now impossible to use the wrong associated session resource with an incorrect session. New code, now looks like:

```rust
use cassandra_cpp::*;

async fn create_session() -> Result<Session> {
    let contact_points = "127.0.0.1";
    let mut cluster = Cluster::default();
    cluster.set_contact_points(contact_points)?;
    cluster.set_load_balance_round_robin();

    cluster.connect().await
}

async fn execute_statement() -> Result<()> {
    let session = create_session().await?;
    let result = session
        .statement("SELECT keyspace_name FROM system_schema.keyspaces;")
        .execute()
        .await?;

    for row in result.iter() {
        let col: String = row.get_by_name("keyspace_name")?;
        print!("ks = {}", col);
    }

    Ok(())
}

async fn execute_prepared_statement() -> Result<()> {
    let session = create_session().await?;
    let prepared_statement = session
        .prepare("SELECT value FROM kv_table WHERE key = ?")
        .await?;

    let mut statement = prepared_statement.bind();
    statement.bind_string(0, "key")?;

    let result = statement.execute().await?;
    for row in result.iter() {
        let col: String = row.get_by_name("value")?;
        print!("value = {}", col);
    }

    Ok(())
}


async fn execute_batch_statement() -> Result<()> {
    let session = create_session().await?;
    let mut batch = session.batch(BatchType::LOGGED);
    // note: binding of statement omitted in this example.
    batch.add_statement(session.statement("INSERT INTO kv_table (key, value) VALUES (?, ?)"))?;
    batch.add_statement(session.statement("INSERT INTO kv_table (key, value) VALUES (?, ?)"))?;
    batch.add_statement(session.statement("INSERT INTO kv_table (key, value) VALUES (?, ?)"))?;
    let _result = batch.execute().await?;

    Ok(())
}
```

There are some pitfalls however, which probably merit additional discussion. `execute()` now consumes the given associated session resource, and thus, if you want to execute the same statement multiple times it must be recreated. Practically, I don't think this is an issue, most code, reasonably, executes a statement once, and if it needs to execute it multiple times, it's because it needs different arguments bound to it. 

I've yet to update any tests/examples/docs, those will come as we gain consensus on whether or not this is an acceptable approach. 

## Potential todo list:
 - [x] Make `CassFuture` private to the crate. All public functions that once returned a `CassFuture` will now just be an `async fn`. 
 - [x] Get rid of `Session::new`, `Session::connect` and `Session::connect_keyspace`, and move those all to `Cluster`. 
 - [x] `Batch::add_statement` should consume statement ownership + check if the session it's bound to matches, now that we have that info. 
 - [X] What to do about `Session::close` now that it's an Arc? 
    - ~~One idea, is that we could use `Arc::try_unwrap` to only allow `Close` to succeed once all associated resources are destroyed.~~
    - ~~The other idea is to leave it as is, lingering associated resources will just start to error once the session is closed.~~
    - Removing for now (see: https://github.com/Metaswitch/cassandra-rs/pull/96#issuecomment-756730646)
 - [x] Get rid of `stmt!` macro. 

## Definite todo list:
  - [x] update tests
  - [x] update examples
  - [ ] update docs
  - [ ] update changelog
  - [ ] migration guide.

### CODE REVIEW

**Comment by jhgg** [comment]
Hey Keith,

I think this is "code complete" - and now we just have to write some docs, migration guide and the changelog. 

PTAL at your earliest convenience.

**Comment by kw217** [comment]
Thanks! I'm out tomorrow but will look at this on Monday.

**Comment by kw217** [comment]
For the record, see our discussion at https://github.com/Metaswitch/cassandra-rs/pull/96 - I'm fully in support of this approach. Thanks!

File: src/cassandra/log.rs [hunk 1]
```diff
@@ -1,7 +1,7 @@
 use crate::cassandra_sys::CassLogLevel_;
 use crate::cassandra_sys::CassLogMessage;
 // use cassandra_sys::cass_log_cleanup; @deprecated
-use crate::cassandra::util::Protected;
+use crate::cassandra::util::ProtectedInner;
```
Comment (kw217) [inline src/cassandra/log.rs:4]: Unused import?

File: tests/basic.rs [hunk 2]
```diff
@@ -542,41 +532,46 @@ fn test_error_reporting() {
...
-fn test_statement_timeout() {
-    let mut query = stmt!("SELECT * FROM system_schema.tables;");
+#[tokio::test]
+async fn test_statement_timeout() {
+    let session = help::create_test_session().await;
+    let mut query = session.statement("SELECT * FROM system_schema.tables;");
     query.set_statement_request_timeout(Some(Duration::milliseconds(30000 as i64)));
-    let session = help::create_test_session();
-    let result = session.execute(&query).wait().unwrap();
+    let result = query.execute().await.unwrap();
```
Comment (kw217) [inline tests/basic.rs:574]: Thanks for sorting all of these tests - the new code looks nice!

**Review by kw217 — CHANGES_REQUESTED** [review]
Thanks - this is looking good so far. A bunch of comments, but only one major one I think: why does `CassFuture` have a `Session`, and does it need to?

GitHub didn't let me comment on a moved file - can you tell me why `schema_meta` example is now disabled?

Also, please can you create a 0.16 migration guide and start making notes in there of how people should change their code?

File: examples/simple.rs [hunk 3]
```diff
@@ -1,24 +1,73 @@
...
+fn assert_send_static_future<F, T>(f: F) -> F
+where
+    F: std::future::Future<Output = T> + Send + 'static,
+{
+    f
+}
+
+#[tokio::main]
+async fn main() -> Result<()> {
+    // tokio, et-al expect the future to be send + 'static, this ensures that the operations we do above are
```
[the remark this answers is not in the document]
Comment (jhgg) [reply examples/simple.rs:65]: Sorry, this belongs in a test and not an example. The purpose is to assert that the future ends up being `Send + 'static`. As it turns out, when you're dealing with FFI, it's very easy to get code that ends up being `!Send` future. This is to statically verify we aren't breaking this guarantee. 

I am going to move this to a test.

File: src/cassandra/session.rs [hunk 4]
```diff
@@ -70,84 +94,53 @@ impl Default for Session {
...
-    //    pub fn execute(&self, statement: &str, parameter_count: u64) -> CassFuture {
-    //        unsafe {
-    //            CassFuture::build(cass_session_execute(self.0,
-    //            Statement::new(statement, parameter_count).inner()))
-    //        }
-    //    }
+    /// Creates a statement with the given query.
+    pub fn statement(&self, query: impl AsRef<str>) -> Statement {
+        let query = query.as_ref();
+        let param_count = query.matches("?").count();
```
[the remark this answers is not in the document]
Comment (jhgg) [reply src/cassandra/session.rs:116]: Sounds good.

File: src/cassandra/statement.rs [hunk 5]
```diff
@@ -67,45 +70,64 @@ use crate::cassandra_sys::CASS_UINT64_MAX;
...
+pub struct Statement(StatementInner, Session);
 
 // The underlying C type has no thread-local state, but does not support access
 // from multiple threads: https://datastax.github.io/cpp-driver/topics/#thread-safety
-unsafe impl Send for Statement {}
+unsafe impl Send for StatementInner {}
 
-impl Protected<*mut _Statement> for Statement {
+impl ProtectedInner<*mut _Statement> for StatementInner {
+    #[inline(always)]
```
[the remark this answers is not in the document]
Comment (jhgg) [reply src/cassandra/statement.rs:98]: I always sprinkle these when I for sure want the compiler to inline things. It probably does anyways, but this is just me being defensive haha.

File: src/cassandra/cluster.rs [hunk 6]
```diff
@@ -150,31 +156,37 @@ impl Cluster {
     }
 
     /// Sets the SSL context and enables SSL
-    pub fn set_ssl(&mut self, ssl: &mut Ssl) -> &Self {
+    pub fn set_ssl(&mut self, ssl: Ssl) -> &Self {
         unsafe {
             cass_cluster_set_ssl(self.0, ssl.inner());
```
[the remark this answers is not in the document]
Comment (jhgg) [reply src/cassandra/cluster.rs:161]: I think this is alright, my analysis below:

```rust
let cluster = cass_cluster_new();
let ssl = cass_ssl_new(); // `ssl`'s ref-count 1
cass_cluster_set_ssl(cluster, ssl); // `ssl`'s ref-count 2 (held by cluster)
cass_ssl_free(ssl); // 'ssl`'s ref-count 1
cass_cluster_free(cluster); // `ssl`'s ref-count 0
```

In the rust code then, this ends up being:

```rust
let mut cluster = Cluster::default();
let ssl = Ssl::default(); // ref-count 1
cluster.set_ssl(ssl); // ref-count 1 (ownership moved into `cluster.set_ssl`, which retains a reference inside of `cluster` (in C++ land), `ssl` however is freed via `Ssl::drop`)
drop(cluster); // ssl ref count 0
```

Am I missing something 🤔

File: src/cassandra/future.rs [hunk 7]
```diff
@@ -37,13 +37,16 @@ use std::task::{Context, Poll, Waker};
...
 #[derive(Debug)]
-pub struct CassFuture<T> {
+pub(crate) struct CassFuture<T> {
     /// The underlying Cassandra driver future object.
     inner: *mut _Future,
 
     /// The current state of this future.
     state: Arc<FutureTarget>,
 
+    /// The session the future is being executed upon.
```
[the remark this answers is not in the document]
Comment (jhgg) [reply src/cassandra/future.rs:47]: :+1: Yeah let me see if I can improve this while keeping the public API the same

**Comment by jhgg** [comment]
Thanks for the review! I'll get to addressing your comments some time this week :tada:

File: examples/simple.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (kw217) [reply examples/simple.rs:65]: Cool, thanks.

File: src/cassandra/statement.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (kw217) [reply src/cassandra/statement.rs:98]: I prefer to trust the compiler, but don't feel strongly. Happy either way.

**Comment by kw217** [comment]
@jhgg are you still working on the API refactor? I'm about to do a major version bump (to require a higher version of the C++ driver) and I'd like to go to 1.0 to reflect the level of support of this driver. If this refactor is close then I can wait, but if not I should go to 1.0 now and we should save it for a future 2.0. What do you think?

**Comment by renancloudwalk** [comment]
@kw217 I think we can move on here. 
Later we can  cherry pick some parts, I can work on this if you all agree. 

But I think it is better to create this Arc interface separated from the "simple" session. I also would try to make the delivery as small and "simple" as possible. Maybe one small change oriented towards having a session Arc with a semaphore.

Breaking changes would not be wise IMO

**Comment by kw217** [comment]
Thanks @renancloudwalk , I'd be very happy if you want to work on this, and I know @jhgg will be fine with that too. Please note there are a bunch of issues and PRs labelled improved-api, not just this one.

If there are parts you think we can adopt without breaking the API, that's great and please go ahead and cherry-pick.

However my sense is that we can't really improve the experience significantly without breaking the API. So - if we're going to break it, let's break it once and make all the fixes we think are needed.

**Comment by renancloudwalk** [comment]
Hi @kw217 I finally took some time to address all the comments you made to @jhgg.
I also added some more documentation, the only thing missing is the migration guide (which I'm having some trouble writing, maybe I can write some draft and we can work together creating a good migration guide)

Shall I open a pull request referencing this one, so we can proceed?

**Comment by kw217** [comment]
Wow, that's great - thanks @renancloudwalk ! Yes please open a new PR and take it from there!

**Comment by kw217** [comment]
This work is now merged under #160. Thanks @jhgg and @renancloudwalk .

**Pull request closed without merging**